### PR TITLE
Restructure ROCM Library Search

### PR DESCRIPTION
**Pull request by dllehr-amd**
Currently there are a handful of ROCM dependant files which are required for triton to run.  The linker(ld.lld), the include files, and multiple hip/hsa shared objects.

This change will provide three search areas to find these files.  All in the same order.

1. $ROCM_PATH environmental variable.  If this exists it will override all other locations to find ROCM necessary files

2. /opt/rocm.  The default location for ROCm installations.  Finding one here will notify triton that ROCM is installed in this environment

3. third_party/rocm.  This location is within the python/triton directory and is carried over when triton is built.  IF all necessary files are in this location there will be no need to have ROCM installed at all on the system.

To ease with step 3.  A new script scripts/amd/setup_rocm_libs.sh has been added to the repo.  Executing this script will cause all necessary ROCM files to be downloaded from their respective packages on repo.radeon.com and installed in third_party/rocm.  Allowing for triton to run without installing the full ROCM stack.  setup_rocm_libs.sh takes a env_var ROCM_VERSION if a user wishes to install a ROCM version other than the default (currently 5.4.2)

When triton whls are built to support Pytorch, method 3 will be used to stay in sync with PyTorch's approach of bringing along any libraries needed and not requiring ROCM to be installed.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -240,9 +240,19 @@ if(TRITON_BUILD_PYTHON_MODULE)
...
 
-  set(ROCM_LIBRARIES
-    ${CMAKE_CURRENT_SOURCE_DIR}/lib/rocm/libhsa-runtime64.so
+  if(EXISTS "$ENV{ROCM_PATH}/lib/libhsa-runtime64.so.1" )
+  set(ROCM_LIBRARIES 
+    $ENV{ROCM_PATH/lib/libhsa-runtime64.so.1}
   )
+  elseif(EXISTS  "/opt/rocm/lib/libhsa-runtime64.so.1" )
+  set(ROCM_LIBRARIES 
+  /opt/rocm/lib/libhsa-runtime64.so.1
```
Comment (jithunnair-amd) [inline CMakeLists.txt:249]: I've been burnt in the past by using a hardcoded string that I thought was invariant, only to have to update all its occurrences at a later point. At least in this file and `lib/Target/HSACO/HSACOTranslation.cpp`, we should be able to use a CMAKE variable for the default value `/opt/rocm` and use it everywhere. And if cmake is invoked in the build by `python/setup.py`, it should be possible to encompass the hardcoded references in `python/triton/compiler.py` as well.

**Review by jithunnair-amd — CHANGES_REQUESTED** [review]
Looks good! A few minor changes requested

**Comment by dllehr-amd** [comment]
Thanks @jithunnair-amd!  Any prints etc will be removed once I'm passing CI.  Your last concern libhsa-runtime.so we can work through, I'm sure there's a better way we can approach this other than hard coding, I wanted to at least get the groundwork in.

**Comment by rsanthanam-amd** [comment]
@micmelesse Please review this PR since it is quite relevant to the upstreaming effort.

File: lib/Target/HSACO/HSACOTranslation.cpp [hunk 2]
```diff
@@ -144,7 +146,33 @@ std::string generate_hsaco(llvm::Module *module, const std::string &triple,
...
-  std::string lld_path = ::triton::tools::getenv("ROCM_PATH") + "/llvm/bin/ld.lld";
+  std::string rocm_path = ::triton::tools::getenv("ROCM_PATH");
+  std::string lld_suffix = "/llvm/bin/ld.lld";
+  std::string lld_path;
+  if (rocm_path.empty()){
+    std::cout << "ROCM_PATH Not found...Trying /opt/rocm" << std::endl;
+    rocm_path = "/opt/rocm";
+  } 
+
+  lld_path = "/opt/rocm" + lld_suffix;
```
Comment (binarman) [inline lib/Target/HSACO/HSACOTranslation.cpp:157]: Maybe you want something like this?
```suggestion
  lld_path = rocm_path + lld_suffix;
```

File: CMakeLists.txt [hunk 3]
```diff
@@ -253,6 +267,7 @@ if(TRITON_BUILD_PYTHON_MODULE)
       ${TRITON_LIBRARIES}
     )
   else()
+  message(HIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII "${ROCM_LIBRARIES}")
```
Comment (jayfurmanek) [inline CMakeLists.txt:270]: LOL.
Need to remove some debug here I think.

File: scripts/amd/setup_rocm_libs.sh [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+
+ROCM_INT=$(($ROCM_VERSION_MAJOR * 10000 + $ROCM_VERSION_MINOR * 100 + $ROCM_VERSION_PATCH))
+OS_NAME=`awk -F= '/^NAME/{print $2}' /etc/os-release`
+
+if [[ $ROCM_INT -eq 50402 ]]; then
+    if [[ "$OS_NAME" == *"CentOS Linux"* ]]; then
+        ROCM_PKGS=(
+            "https://repo.radeon.com/rocm/rhel8/5.4.2/main/hsa-rocr5.4.2-1.7.0.50402-104.el8.x86_64.rpm"
+            "https://repo.radeon.com/rocm/rhel8/5.4.2/main/hip-runtime-amd-5.4.22803.50402-104.el8.x86_64.rpm"
+            "https://repo.radeon.com/rocm/rhel8/5.4.2/main/hip-devel-5.4.22803.50402-104.el8.x86_64.rpm"
```
Comment (micmelesse) [inline scripts/amd/setup_rocm_libs.sh:42]: This is pinned to a specific rocm version. We should make this a general soltion.

**Review by micmelesse — CHANGES_REQUESTED** [review]

**Comment by jayfurmanek** [comment]
This one needs refreshed to match: https://github.com/ROCmSoftwarePlatform/triton/commit/e6aea90fb3e8218cb562e5d990719112d8282702

**Comment by micmelesse** [comment]
@dllehr-amd Can you update this branch to resolve conflicts?

File: scripts/amd/setup_rocm_libs.sh (hunk 4 above)
Comment (jataylo) [reply scripts/amd/setup_rocm_libs.sh:42]: This is now a general solution and will pull the libraries from any manylinux image for building the wheel https://github.com/ROCmSoftwarePlatform/triton/pull/354

**Comment by jataylo** [comment]
Change came in via https://github.com/ROCmSoftwarePlatform/triton/pull/354 instead

**Pull request closed without merging**